lms/djangoapps/courseware/model_data.py

```diff
@@ -34,12 +36,644 @@ class InvalidWriteError(Exception):
     """
 
 
-def chunks(items, chunk_size):
+def _all_usage_keys(descriptors, aside_types):
     """
-    Yields the values from items in chunks of size chunk_size
+    Return a set of all usage_ids for the `descriptors` and for
+    as all asides in `aside_types` for those descriptors.
     """
-    items = list(items)
-    return (items[i:i + chunk_size] for i in xrange(0, len(items), chunk_size))
+    usage_ids = set()
+    for descriptor in descriptors:
+        usage_ids.add(descriptor.scope_ids.usage_id)
+
+        for aside_type in aside_types:
+            usage_ids.add(AsideUsageKeyV1(descriptor.scope_ids.usage_id, aside_type))
+
+    return usage_ids
+
+
+def _all_block_types(descriptors, aside_types):
+    """
+    Return a set of all block_types for the supplied `descriptors` and for
+    the asides types in `aside_types` associated with those descriptors.
+    """
+    block_types = set()
+    for descriptor in descriptors:
+        block_types.add(BlockTypeKeyV1(descriptor.entry_point, descriptor.scope_ids.block_type))
+
+    for aside_type in aside_types:
+        block_types.add(BlockTypeKeyV1(XBlockAside.entry_point, aside_type))
+
+    return block_types
+
+
+class DjangoKeyValueStore(KeyValueStore):
+    """
+    This KeyValueStore will read and write data in the following scopes to django models
+        Scope.user_state_summary
+        Scope.user_state
+        Scope.preferences
+        Scope.user_info
+
+    Access to any other scopes will raise an InvalidScopeError
+
+    Data for Scope.user_state is stored as StudentModule objects via the django orm.
+
+    Data for the other scopes is stored in individual objects that are named for the
+    scope involved and have the field name as a key
+
+    If the key isn't found in the expected table during a read or a delete, then a KeyError will be raised
+    """
+
+    _allowed_scopes = (
+        Scope.user_state_summary,
+        Scope.user_state,
+        Scope.preferences,
+        Scope.user_info,
+    )
+
+    def __init__(self, field_data_cache):
+        self._field_data_cache = field_data_cache
+
+    def get(self, key):
+        self._raise_unless_scope_is_allowed(key)
+        return self._field_data_cache.get(key)
+
```

Comment:
Yes, and it exists in XBlockUserStateClient. But it's not in XBlock `field_data` yet, and I don't want to add it as part of this story.